File: Sources/Controllers/Misc/Views/PopoverVC.m
#import "PopoverVC.h"


@implementation PopoverVC {

	UILabel *infoLabel;

}


- (void)viewDidLoad {

	[super viewDidLoad];

	// do any additional setup after loading the view, typically from a nib.
	UIColor *firstColor = kAzureMintTintColor;
	UIColor *secondColor = [UIColor colorWithRed:0.40 green:0.81 blue:0.78 alpha: 1.0];
	NSArray *gradientColors = @[(id)firstColor.CGColor, (id)secondColor.CGColor];

	CAGradientLayer *gradientLayer = [CAGradientLayer layer];
	gradientLayer.colors = gradientColors;
	gradientLayer.frame = self.view.bounds;
	gradientLayer.startPoint = CGPointMake(0, 0);
	gradientLayer.endPoint = CGPointMake(1, 1);
	[self.view.layer addSublayer: gradientLayer];

	infoLabel = [UILabel new];
	infoLabel.font = [UIFont systemFontOfSize: 12];
	infoLabel.alpha = 0;
	infoLabel.transform = CGAffineTransformMakeScale(0.1, 0.1);
	infoLabel.numberOfLines = 0;
	infoLabel.textAlignment = NSTextAlignmentCenter;
	infoLabel.adjustsFontSizeToFitWidth = YES;
	infoLabel.translatesAutoresizingMaskIntoConstraints = NO;
	[self.view addSubview: infoLabel];

	[infoLabel.topAnchor constraintEqualToAnchor: self.view.safeAreaLayoutGuide.topAnchor constant: 5].active = YES;
	[infoLabel.leadingAnchor constraintEqualToAnchor: self.view.leadingAnchor constant: 5].active = YES;
	[infoLabel.trailingAnchor constraintEqualToAnchor: self.view.trailingAnchor constant: -5].active = YES;
	[infoLabel.centerXAnchor constraintEqualToAnchor: self.view.centerXAnchor].active = YES;

}


- (void)fadeInPopoverWithMessage:(NSString *)message {

	infoLabel.text = message;

}


- (void)viewWillAppear:(BOOL)animated {

	[super viewWillAppear: animated];

	self.popoverPresentationController.containerView.alpha = 0;
	self.popoverPresentationController.containerView.transform = CGAffineTransformMakeScale(0.1, 0.1);

	dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 2.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{

		[UIView animateWithDuration:0.5 delay:0 options:UIViewAnimationOptionTransitionCrossDissolve animations:^{

			self.popoverPresentationController.containerView.alpha = 1;
			self.popoverPresentationController.containerView.transform = CGAffineTransformMakeScale(1, 1);

		} completion:^(BOOL finished) {

			[UIView animateWithDuration:0.5 delay:0.5 usingSpringWithDamping:0.6 initialSpringVelocity:0.2 options:UIViewAnimationOptionTransitionCrossDissolve animations:^{

				infoLabel.alpha = 1;
				infoLabel.transform = CGAffineTransformMakeScale(1, 1);

			} completion:^(BOOL finished) {

				dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 3.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{

					[self dismissViewControllerAnimated:YES completion:nil];

				});

			}];

		}];

	});

}

@end

File: Sources/Controllers/Misc/Views/PopoverVC.swift
import UIKit

/// Controller that'll show a view as a popover
final class PopoverVC: UIViewController {

	private lazy var gradientLayer: CAGradientLayer = {
		let firstColor = UIColor.kAzureMintTintColor
		let secondColor = UIColor(red: 0.40, green: 0.81, blue: 0.78, alpha: 1.0)
		let gradientColors = [firstColor.cgColor, secondColor.cgColor]
		let layer = CAGradientLayer()
		layer.colors = gradientColors
		layer.frame = view.bounds
		layer.startPoint = CGPoint(x: 0, y: 0)
		layer.endPoint = CGPoint(x: 1, y: 1)
		return layer
	}()

	private lazy var infoLabel: UILabel = {
		let label = UILabel()
		label.font = .systemFont(ofSize: 12)
		label.alpha = 0
		label.transform = .init(scaleX: 0.1, y: 0.1)
		label.numberOfLines = 0
		label.textAlignment = .center
		label.adjustsFontSizeToFitWidth = true
		label.translatesAutoresizingMaskIntoConstraints = false
		view.addSubview(label)
		return label
	}()

	// ! Lifecycle

	override func viewDidLoad() {
		super.viewDidLoad()

		view.layer.insertSublayer(gradientLayer, at: 0)

		infoLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 5).isActive = true
		infoLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 5).isActive = true
		infoLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -5).isActive = true
		infoLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
	}

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)

		popoverPresentationController?.containerView?.alpha = 0
		popoverPresentationController?.containerView?.transform = .init(scaleX: 0.1, y: 0.1)

		DispatchQueue.main.asyncAfter(deadline: .now() + 2.5) {
			UIView.animate(withDuration: 0.5, animations: {
				self.popoverPresentationController?.containerView?.alpha = 1
				self.popoverPresentationController?.containerView?.transform = .init(scaleX: 1, y: 1)
			}) { _ in
				UIView.animate(withDuration: 0.5, delay: 0.5, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.2, animations: {
					self.infoLabel.alpha = 1
					self.infoLabel.transform = .init(scaleX: 1, y: 1)
				}) { _ in
					DispatchQueue.main.asyncAfter(deadline: .now() + 3.5) {
						self.dismiss(animated: true)
					}
				}
			}
		}
	}

}

extension PopoverVC {

	// ! Public

	/// Function to show fade in a popover with a given message
	/// - Parameters:
	///		- withMessage: A string representing the message
	func fadeInPopover(withMessage message: String) { infoLabel.text = message }

}
